test/testutil.c

```diff
@@ -145,18 +145,261 @@ int run_tests(const char *test_prog_name)
+    DEFINE_COMPARISON(type, name, ne, !=, fmt)                          \
+    DEFINE_COMPARISON(type, name, lt, <, fmt)                           \
+    DEFINE_COMPARISON(type, name, le, <=, fmt)                          \
+    DEFINE_COMPARISON(type, name, gt, >, fmt)                           \
+    DEFINE_COMPARISON(type, name, ge, >=, fmt)
+
+DEFINE_COMPARISONS(int, int, "%d")
+DEFINE_COMPARISONS(unsigned int, uint, "%u")
+DEFINE_COMPARISONS(char, char, "%c")
+DEFINE_COMPARISONS(unsigned char, uchar, "%u")
+DEFINE_COMPARISONS(long, long, "%ld")
+DEFINE_COMPARISONS(unsigned long, ulong, "%lu")
+DEFINE_COMPARISONS(size_t, size_t, "%" OSSLzu)
+
+DEFINE_COMPARISON(void *, ptr, eq, ==, "%p")
+DEFINE_COMPARISON(void *, ptr, ne, !=, "%p")
+
+int test_ptr_null(const void *p, const char *desc, ...)
+{
+    va_list ap;
+
+    if (p == NULL)
+        return 1;
+    va_start(ap, desc);
+    test_fail_message(1, desc, ap, "ptr", "%p == NULL", p);
+    va_end(ap);
+    return 0;
+}
+
+int test_ptr(const void *p, const char *desc, ...)
+{
+    va_list ap;
+
+    if (p != NULL)
+        return 1;
+    va_start(ap, desc);
+    test_fail_message(1, desc, ap, "ptr", "%p != NULL", p);
+    va_end(ap);
+    return 0;
+}
+
+int test_true(int b, const char *desc, ...)
+{
+    va_list ap;
+
+    if (b)
+        return 1;
+    va_start(ap, desc);
+    test_fail_message(1, desc, ap, "bool", "false == true");
+    va_end(ap);
+    return 0;
+}
+
+int test_false(int b, const char *desc, ...)
+{
+    va_list ap;
+
+    if (!b)
+        return 1;
+    va_start(ap, desc);
+    test_fail_message(1, desc, ap, "bool", "true == false");
+    va_end(ap);
+    return 0;
+}
+
 static const char *print_string_maybe_null(const char *s)
 {
     return s == NULL ? "(NULL)" : s;
 }
 
-int strings_equal(const char *desc, const char *s1, const char *s2)
+int test_str_eq(const char *s1, const char *s2, const char *desc, ...)
 {
+    va_list ap;
+
     if (s1 == NULL && s2 == NULL)
       return 1;
     if (s1 == NULL || s2 == NULL || strcmp(s1, s2) != 0) {
-        fprintf(stderr, "%s mismatch: %s vs %s\n", desc, print_string_maybe_null(s1),
-                print_string_maybe_null(s2));
+        va_start(ap, desc);
+        test_fail_message(1, desc, ap, "string", "%s == %s",
+                          print_string_maybe_null(s1), print_string_maybe_null(s2));
+        va_end(ap);
+        return 0;
+    }
+    return 1;
+}
+
+int test_str_ne(const char *s1, const char *s2, const char *desc, ...)
+{
+    va_list ap;
+
+    if ((s1 == NULL) ^ (s2 == NULL))
+      return 1;
+    if (s1 == NULL || strcmp(s1, s2) == 0) {
+        va_start(ap, desc);
+        test_fail_message(1, desc, ap, "string", "%s != %s",
+                          print_string_maybe_null(s1), print_string_maybe_null(s2));
+        va_end(ap);
+        return 0;
+    }
+    return 1;
+}
+
+
+int test_mem_eq(const void *s1, const void *s2, size_t n, const char *desc, ...)
```

Comment:
You could consider doing `test_mem_eq(const void *s1, size_t s1_len, const void *s2, size_t s2_len)`, then tests don't have to check the length separately.